imagetest/test_suites/oslogin/ssh_test.go

```diff
@@ -271,3 +276,192 @@ func TestAdminSSH(t *testing.T) {
+	privateSSHKey, err := utils.AccessSecret(ctx, secretClient, normal2FASSHKey)
+	if err != nil {
+		t.Fatalf("failed to get user key: %v", err)
+	}
+
+	// Manually set up SSH.
+	vmname, err := utils.GetRealVMName("twofa")
+	if err != nil {
+		t.Fatalf("failed to get hostname: %v", err)
+	}
+	hostname := fmt.Sprintf("%s:22", vmname)
+	signer, err := ssh.ParsePrivateKey([]byte(privateSSHKey))
+	if err != nil {
+		t.Fatalf("failed to parse private key: %v", err)
+	}
+
+	cb := func(name, instruction string, questions []string, echos []bool) (answers []string, err error) {
+		if len(questions) == 0 {
+			return
+		}
+
+		answers = make([]string, 1)
+		firstQuestionRegex, err := regexp.Compile(".*Enter.*number.*")
+		if err != nil {
+			return
+		}
+		if firstQuestionRegex.MatchString(questions[0]) {
+			// 1 is the response for Authenticator OTP
+			answers[0] = "1"
+			return
+		}
+
+		// If not the first question, input code for two-factor.
+		ctx := context.Background()
+		secretClient, err := secretmanager.NewClient(ctx)
+		if err != nil {
+			return
+		}
+		defer secretClient.Close()
+
+		s, err := utils.AccessSecret(ctx, secretClient, normal2FAKey)
+		if err != nil {
+			return
+		}
+		secret := strings.ToUpper(s)
+
+		if !gotp.IsSecretValid(secret) {
+			// Avoid panic that doesn't return a useful error message for test runner
+			err = fmt.Errorf("invalid secret")
+			return
+		}
+		code := gotp.NewDefaultTOTP(secret).Now()
+		answers[0] = code
+		return
+	}
+
+	sshConfig := &ssh.ClientConfig{
+		User:            posix,
+		Auth:            []ssh.AuthMethod{ssh.PublicKeys(signer), ssh.KeyboardInteractive(cb)},
+		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
+	}
+
+	client, err := ssh.Dial("tcp", hostname, sshConfig)
+	if err != nil {
+		t.Fatalf("failed to create ssh client: %v", err)
+	}
+	defer client.Close()
+
+	if err = sessionOSLoginEnabled(client); err != nil {
+		t.Fatalf("%v", err)
+	}
+}
+
+func Test2FAAdminSSH(t *testing.T) {
+	ctx := utils.Context(t)
+
+	// Secret manager client.
+	secretClient, err := secretmanager.NewClient(ctx)
+	if err != nil {
+		t.Fatalf("failed to create secret manager client: %v", err)
+	}
+	defer secretClient.Close()
+
+	user, err := utils.AccessSecret(ctx, secretClient, admin2FAUser)
+	if err != nil {
+		t.Fatalf("failed to get user info: %v", err)
+	}
+	posix := getPosix(user)
+
+	privateSSHKey, err := utils.AccessSecret(ctx, secretClient, admin2FASSHKey)
+	if err != nil {
+		t.Fatalf("failed to get user key: %v", err)
+	}
+
+	// Manually set up SSH.
+	vmname, err := utils.GetRealVMName("twofa")
+	if err != nil {
+		t.Fatalf("failed to get hostname: %v", err)
+	}
+	hostname := fmt.Sprintf("%s:22", vmname)
+	signer, err := ssh.ParsePrivateKey([]byte(privateSSHKey))
+	if err != nil {
+		t.Fatalf("failed to parse private key: %v", err)
+	}
+
+	// Callback function for SSH keyboard-interactive challenges.
+	cb := func(name, instruction string, questions []string, echos []bool) (answers []string, err error) {
```

Comment:
Can we generalize this callback? as it's repeating itself here and in the other test function.